### ISSUE TITLE
False positive URF_UNREAD_FIELD with spotbugs 4.7.2

### ISSUE THREAD

**Issue by trancexpress** [spotbugs/spotbugs#2174]
Start Eclipse in a clean workspace, have `spotbugs` 4.7.2 in the installation.

1. Import project from:  [TestSpotBugs.zip](https://github.com/spotbugs/spotbugs/files/9547793/TestSpotBugs.zip)
2. Open type `test.Test`.
3. Right-click in the editor -> `SpotBugs` submenu -> `Find Bugs`
4. Observe unexpected warning:
```
    Bug: Unread field: test.Test.object
    This field is never read.  Consider removing it from the class.

    Rank: Of Concern (18), confidence: Normal
    Pattern: URF_UNREAD_FIELD 
    Type: UrF, Category: PERFORMANCE (Performance)
```

This is a regression from: https://github.com/spotbugs/spotbugs/commit/4c9b635d8de26e691add3f291f980ad21cf5abd1

In particular, applying this partial revert fixes the problem:

```
diff --git a/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java b/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java
index 3f932e5a2..1336f97c4 100644
--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java
@@ -2657,35 +2657,13 @@ public class OpcodeStack {
             Item result = new Item(JAVA_UTIL_ARRAYS_ARRAY_LIST);
             push(result);
             return;
-        } else if (seen == Const.INVOKESTATIC) {
-            Item requestParameter = null;
-            if ("(Ljava/util/List;)Ljava/util/List;".equals(signature)
-                    && "java/util/Collections".equals(clsName)) {
-                requestParameter = top();
-            }
-            String returnTypeName = IMMUTABLE_RETURNER_MAP.get(Pair.of(clsName, methodName));
-            if (returnTypeName != null) {
-                SignatureParser sp = new SignatureParser(signature);
-                for (int i = 0; i < sp.getNumParameters(); ++i) {
-                    pop();
-                }
-                Item result;
-                if (requestParameter != null && JAVA_UTIL_ARRAYS_ARRAY_LIST.equals(requestParameter.getSignature())) {
-                    result = new Item("Ljava/util/Collections$UnmodifiableRandomAccessList");
-                } else {
-                    result = new Item(returnTypeName);
-                }
-                push(result);
-                return;
-            } else if (requestParameter != null) {
-                pop();
-                if (JAVA_UTIL_ARRAYS_ARRAY_LIST.equals(requestParameter.getSignature())) {
-                    Item result = new Item(JAVA_UTIL_ARRAYS_ARRAY_LIST);
-                    push(result);
-                    return;
-                }
-                push(requestParameter); // fall back to standard logic
+        } else if (seen == Const.INVOKESTATIC && "(Ljava/util/List;)Ljava/util/List;".equals(signature)
+                && "java/util/Collections".equals(clsName)) {
+            Item requestParameter = pop();
+            if (JAVA_UTIL_ARRAYS_ARRAY_LIST.equals(requestParameter.getSignature())) {
+                Item result = new Item(JAVA_UTIL_ARRAYS_ARRAY_LIST);
             }
+            push(requestParameter); // fall back to standard logic
         }
 
         pushByInvoke(dbc, seen != Const.INVOKESTATIC);
@@ -2845,7 +2823,7 @@ public class OpcodeStack {
                 Item newValue = mergeFrom.get(i);
                 mergeInto.add(newValue);
                 changed = true;
-            
+
             }
              */
             if (DEBUG2 && changed) {

```

**Comment by baloghadamsoftware** [comment]
```
if (requestParameter != null && JAVA_UTIL_ARRAYS_ARRAY_LIST.equals(requestParameter.getSignature())) {
        result = new Item("Ljava/util/Collections$UnmodifiableRandomAccessList");
```

This is a bug for sure. A semicolon is missing from the end of the signature. It is strange that all the tests (including all the tests on open-source projects in our CI job) missed this.

**Comment by iloveeclipse** [comment]
> This is a bug for sure.

@baloghadamsoftware : would you like to provide a fix for this regression?

**Comment by baloghadamsoftware** [comment]
> > This is a bug for sure.
> 
> @baloghadamsoftware : would you like to provide a fix for this regression?

Yes, of course, it is my fault, I will provide a fix today.

**Comment by trancexpress** [comment]
I can confirm, this does fix the problem:

```
diff --git a/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java b/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java
index 3f932e5a2..73dfba89f 100644
--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java
@@ -2671,7 +2671,7 @@ public class OpcodeStack {
                 }
                 Item result;
                 if (requestParameter != null && JAVA_UTIL_ARRAYS_ARRAY_LIST.equals(requestParameter.getSignature())) {
-                    result = new Item("Ljava/util/Collections$UnmodifiableRandomAccessList");
+                    result = new Item("Ljava/util/Collections$UnmodifiableRandomAccessList;");
                 } else {
                     result = new Item(returnTypeName);
                 }
```

Both in the reproducer and in the original case (that cannot be shared due to IP).

**Comment by iloveeclipse** [comment]
Fixed via #2175
